### PR TITLE
fix(ssh-keys-sidebar): SSH keys sidebar loads invalid key if there ar…

### PR DESCRIPTION
**Pull request by HeyRoach**
…e duplicate names for different accounts

### CODE REVIEW

File: src/app/ssh-keys/ssh-key-list/ssh-key-list.component.ts [hunk 1]
```diff
@@ -40,7 +40,7 @@ export class SshKeyListComponent {
   public selectSshKeyPair(sshKeyPair: SSHKeyPair): void {
     this.router.navigate(['view', sshKeyPair.name], {
       relativeTo: this.route,
-      queryParamsHandling: 'preserve'
+      queryParams: {account: sshKeyPair.account}
```
Comment (zolotyx) [inline src/app/ssh-keys/ssh-key-list/ssh-key-list.component.ts]: Please, add queryParamsHandling: 'preserve'

File: src/app/shared/services/ssh-keypair.service.ts [hunk 2]
```diff
@@ -20,15 +21,23 @@ export interface SshKeyCreationData {
...
     super(http);
+
+    this.activatedRoute.queryParams.subscribe((params: Params) => {
+      this.account = params['account'];
+    });
   }
 
   public getByName(name: string): Observable<SSHKeyPair> {
-    return this.getList({ name }).map(sshKeys => sshKeys[0]);
+    return this.getList({ name }).map(sshKeys => {
```
Comment (zolotyx) [inline src/app/shared/services/ssh-keypair.service.ts]: is it possible to fetch only one key by account **and** name? there will be no need to do subscription in constructor and store it locally

Comment (zolotyx) [reply src/app/shared/services/ssh-keypair.service.ts]: You also have to handle the situation when account is not present in the query string

File: src/app/ssh-keys/ssh-key-list/ssh-key-list.component.ts (hunk 1 above)
Comment (HeyRoach) [reply src/app/ssh-keys/ssh-key-list/ssh-key-list.component.ts]: if queryParams and queryParamsHandling: 'preserve' are used together, the query parameters are not adding to the routerLink

Comment (zolotyx) [reply src/app/ssh-keys/ssh-key-list/ssh-key-list.component.ts]: @HeyRoach , is there any other way to pass params to route with queryParamsHandling: 'preserve'?

File: src/app/shared/services/ssh-keypair.service.ts [hunk 3]
```diff
@@ -20,15 +21,21 @@ export interface SshKeyCreationData {
...
     super(http);
+
+    this.activatedRoute.queryParams.subscribe((params: Params) => {
+      this.account = params['account'];
+    });
   }
 
-  public getByName(name: string): Observable<SSHKeyPair> {
-    return this.getList({ name }).map(sshKeys => sshKeys[0]);
+  public getByName(name: string, account?: string): Observable<SSHKeyPair> {
```
Comment (zolotyx) [inline src/app/shared/services/ssh-keypair.service.ts]: refactor this method to accept params object, rename the method

File: src/app/shared/services/ssh-keypair.service.ts [hunk 4]
```diff
@@ -20,15 +21,21 @@ export interface SshKeyCreationData {
   entityModel: SSHKeyPair
 })
 export class SSHKeyPairService extends BaseBackendCachedService<SSHKeyPair> {
-  constructor(
-    private asyncJobService: AsyncJobService,
-    protected http: HttpClient
-  ) {
+  protected account: string;
```
Comment (zolotyx) [inline src/app/shared/services/ssh-keypair.service.ts]: why do you need to store the account? service should not store any data

File: src/app/shared/services/ssh-keypair.service.ts [hunk 5]
```diff
@@ -20,15 +21,21 @@ export interface SshKeyCreationData {
...
+
+  constructor(private asyncJobService: AsyncJobService,
+              protected http: HttpClient,
+              protected activatedRoute: ActivatedRoute) {
     super(http);
+
+    this.activatedRoute.queryParams.subscribe((params: Params) => {
```
Comment (zolotyx) [inline src/app/shared/services/ssh-keypair.service.ts]: remove this

File: src/app/ssh-keys/ssh-key-sidebar/ssh-key-sidebar.component.ts [hunk 6]
```diff
@@ -15,24 +15,31 @@ import { EntityDoesNotExistError } from '../../shared/components/sidebar/entity-
...
+  constructor(protected entityService: SSHKeyPairService,
+              protected notificationService: NotificationService,
+              protected route: ActivatedRoute,
+              protected router: Router,
+              protected userTagService: UserTagService) {
     super(entityService, notificationService, route, router);
   }
 
+  public ngOnInit() {
+    this.getEntityAccount();
```
Comment (zolotyx) [inline src/app/ssh-keys/ssh-key-sidebar/ssh-key-sidebar.component.ts]: remove this, use RXJS power not to store the data in the component

File: src/app/ssh-keys/ssh-key-sidebar/ssh-key-sidebar.component.ts [hunk 7]
```diff
@@ -15,24 +15,31 @@ import { EntityDoesNotExistError } from '../../shared/components/sidebar/entity-
...
+  }
+
   public onDescriptionChange(description: string): void {
     this.description = description;
     this.userTagService.setSshKeyDescription(this.entity, this.description).subscribe();
   }
 
-  protected loadEntity(name: string): Observable<SSHKeyPair> {
-    return this.entityService.getByName(name)
+  protected loadEntity(name: string, account?: string): Observable<SSHKeyPair> {
```
Comment (zolotyx) [inline src/app/ssh-keys/ssh-key-sidebar/ssh-key-sidebar.component.ts]: can we leave this method with a single parameter and fetch account from stateParams here?

File: src/app/shared/services/ssh-keypair.service.ts [hunk 8]
```diff
@@ -1,5 +1,6 @@
 import { HttpClient } from '@angular/common/http';
 import { Injectable } from '@angular/core';
+import { ActivatedRoute } from '@angular/router';
```
Comment (zolotyx) [inline src/app/shared/services/ssh-keypair.service.ts]: remove this

File: src/app/shared/services/ssh-keypair.service.ts [hunk 9]
```diff
@@ -20,15 +21,13 @@ export interface SshKeyCreationData {
   entityModel: SSHKeyPair
 })
 export class SSHKeyPairService extends BaseBackendCachedService<SSHKeyPair> {
-  constructor(
-    private asyncJobService: AsyncJobService,
-    protected http: HttpClient
-  ) {
+  constructor(private asyncJobService: AsyncJobService,
```
Comment (zolotyx) [inline src/app/shared/services/ssh-keypair.service.ts]: revert the indent

File: src/app/ssh-keys/ssh-key-sidebar/ssh-key-sidebar.component.ts [hunk 10]
```diff
@@ -13,16 +13,15 @@ import { EntityDoesNotExistError } from '../../shared/components/sidebar/entity-
...
+  public account: string;
 
-  constructor(
-    protected entityService: SSHKeyPairService,
-    protected notificationService: NotificationService,
-    protected route: ActivatedRoute,
-    protected router: Router,
-    protected userTagService: UserTagService
-  ) {
+  constructor(protected entityService: SSHKeyPairService,
```
Comment (zolotyx) [inline src/app/ssh-keys/ssh-key-sidebar/ssh-key-sidebar.component.ts]: use correct styling for constructor

File: src/app/ssh-keys/ssh-key-sidebar/ssh-key-sidebar.component.ts [hunk 11]
```diff
@@ -13,16 +13,15 @@ import { EntityDoesNotExistError } from '../../shared/components/sidebar/entity-
   selector: 'cs-ssh-key-sidebar',
   templateUrl: 'ssh-key-sidebar.component.html'
 })
-export class SshKeySidebarComponent extends SidebarComponent<SSHKeyPair> implements OnInit {
+export class SshKeySidebarComponent extends SidebarComponent<SSHKeyPair> {
   public description: string;
+  public account: string;
```
Comment (zolotyx) [inline src/app/ssh-keys/ssh-key-sidebar/ssh-key-sidebar.component.ts]: do not store account

File: src/app/ssh-keys/ssh-key-sidebar/ssh-key-sidebar.component.ts [hunk 12]
```diff
@@ -32,7 +31,10 @@ export class SshKeySidebarComponent extends SidebarComponent<SSHKeyPair> impleme
   }
 
   protected loadEntity(name: string): Observable<SSHKeyPair> {
-    return this.entityService.getByName(name)
+    const account = this.route.snapshot.queryParams['account'];
```
Comment (zolotyx) [inline src/app/ssh-keys/ssh-key-sidebar/ssh-key-sidebar.component.ts]: do not use snapshot, use switchMap with this.entityService.getByParams

**Pull request merged**